Remake the project with C# and Avalonia

Rosentti opened the pull request:
# Why

While C++ and Qt is really performant, it's not the most easy thing to use and the existing jungle of spaghetti sucks and I don't really want to have to deal with C++ again for every aspect of the project.

The initial "release" was basically only ever meant to be a testing ground for seeing if it's possible to make a Steam GUI ourself, but because of the initial positive feedback I want to keep maintaining this project, which isn't possible currently.  

Rewriting this in C# should make this a client that the community can integrate cool things into, like ProtonDB support, theming, game issue workarounds, easy custom Proton version downloading and a whole lot else. 

# Problems with C++ and the codebase:
- It makes my life much harder (C++ is HARD)
- It makes it harder to get contributors (C++ is HARD and the spaghetti...)
- It makes system requirements pretty high (Qt6 is NEW)
- Cross-platform isn't as easy (In C#/Avalonia it comes OOTB basically)
- Sharing code kind of sucks
- UI is dreadful to write
- Debugging is a pain, especially for the average Joe sending in a crash report

# With C# and Avalonia:
- MVVM pattern makes it nice and easy to make UI
- Avalonia's default style looks fantastic on all platforms
- Code sharing is very easy
- C# has all the goodies built in like string formatting
- C# is much easier so getting contributors should be a lot easier
- Don't have to worry about memory corruption and race conditions, only when dealing directly with natives

duarm commented:
> UI is dreadful to write

have you tried qml instead of hand-writing the ui?

I'm all for ditching c++ and qt, but have you considered a binding to qt?  As far as I know, only the rust and the python binding are not abandoned, and I wouldn't like a python program.

SFort commented:
I know nothing about Avalonia, and am indiffrent to c++/c# in general.
But if you looking for opinions, i'd say choosing what's simplest for you is a good idea (persumably c#).

A thing that tends to worry me, is having lots of unused deps. (sometimes running and unused).
But considering the initial test release already had build flags, i'm not worried about it.

> I'm all for ditching c++ and qt, but have you considered a binding to qt? As far as I know, only the rust and the python binding are not abandoned, and I wouldn't like a python program.

Can second the anti-python sentiment.
But i also don't particularly care about rust and qt even though i do like durams color scheme,
although i assume something simmilar can be done with Avalonia?

Rosentti commented:
Yeah, QML could've been pretty good, however it is just JS, without the added benefits of something like electron.

pchome commented:
I expected to see something like `steam://open/minigameslist` here, something lightweight with offline support to launch some installed singleplayer games. Still looking into sources if I can adapt it to be like this.

I have some rhetoric questions because of this remake:

* Why a fancy ui? Or clone steam client ui?
  Regular Qt app ui should do the trick (menu, panel, status bar, standard theme).
* C# and "getting contributors" on linux?
  Most people will sit on release builds, others will need to setup C# dev env.

I'd better chose Python for remake (like a tool, not because I like it or know)
* there are already some libs to interact with steam and its files
* simple enough for most linux gamers to understand and contribute
* available almost everywhere (we need it for Proton)

See (for example):
https://wiki.qt.io/PySide2 / https://wiki.qt.io/PySide6  - Qt gui
https://github.com/ValvePython/steam - should do most of the work (not Valve project, just named like this)
https://github.com/ValvePython/steamctl - cli (w/o launcher)

Rosentti commented:
The old codebase was basically just an initial test hacked together fast. Still the original source code can be used by someone smarter than me to potentially code a native Qt/GTK app. C# is also great since it has SteamKit2 and DepotDownloader, so we could potentially change download and verify code ourself. Also I want to make a client I can daily drive, and that leaves no room for bugs. And I also want the client to look better, maybe even clone the new UI somewhat, just more performantly.

a1batross commented:
While C# is a better choice to not deal with UB, raw memory and threads, it's still mostly controlled by a single entity (Microsoft) and to me was always a pain to run anything. Development tools for C# on Linux are varying from non-existent to barely functioning.

Sorry, didn't want to sound like a fossbro, but C# is still a concern.

murlakatamenka commented:
Just for the sake of it, I'll mention such an option:

- core in Rust
- GUI via
  - GTK ([Rust bindings](https://gtk-rs.org/) are in good shape, there is also [relm4](https://crates.io/crates/relm4) built on top of those bindings)
  - Qt5 or Qt6 via PySide (official Python bindings for Qt)

Yeah, can't say this is likely, but Rust is great with its safety and performance / fearless concurrency and C interop if needed etc. while GTK/Qt are classical GUI choices on Linux.

duarm commented:
> While C# is a better choice to not deal with UB, raw memory and threads, it's still mostly controlled by a single entity (Microsoft) and to me was always a pain to run anything. Development tools for C# on Linux are varying from non-existent to barely functioning.

Agree completely, I don't like GTK for the same reasons, and the fact it loves to start a bunch of services with no way to disable (dconf, at-spi).

If the dev is confortable with avalonia, he should use it, but separating the core would help alternate front-ends a lot.

a1batross commented:
Something tells it's going to be a heated discussion.

Rosentti commented:
I didn't pick python as I've never really liked it and (sorry to all devs I may offend) they always seem to have issues. I absolutely love how convenient bottles is, but sometimes it just breaks (not an issue with bottles itself, rather some packaging issue?) and it requires a ton of AUR dependencies. As for some of your concerns:
- The code is split into OpenSteamworks, Common, ClientUI and ClientConsole. Common aims to implement bootstrapping and most of the client code, where ClientUI and ClientConsole just essentially implement "ViewModels" style things
- I develop the client solely on linux, but I will test it on Windows as well, so yes good editors exist (I use VSCode)
- Tooling so far has been fantastic to work with

Rosentti committed:
how'd that comma slip in there?

Rosentti committed:
nope

Rosentti committed:
try adding artifacts support

Rosentti committed:
maybe fix actions

Rosentti committed:
gh why

Rosentti committed:
almost....

Rosentti committed:
incomplete translation system

Rosentti committed:
- translation system
- config for SteamHTML and SteamService
- allow rethrow in LoadWithOrCreate
- AssemblyFolder in ConfigManager
- LoginUsers in ConfigManager
- reduced log spam of htmlhost_fakepid
- json converter for CSteamID

Rosentti committed:
large refactor, quitting is broken for now...

Rosentti committed:
WIP client factory work

Rosentti committed:
container rewrite finished

Rosentti committed:
fixed some warnings

Rosentti committed:
misc

Rosentti committed:
placeholder icons
account picker window TODO

Rosentti committed:
basic logon logic

Rosentti committed:
architected some more login stuffs

Rosentti committed:
logging in to existing account works
broke "add account" though

Rosentti committed:
commit in progress work

Rosentti committed:
finished up ProtoMsg. Doesn't compile yet...

Rosentti committed:
- qr login
- code cleanups
- sharedconnection
- cutlbuffer fixups

Rosentti committed:
fix freezing

Rosentti committed:
code comments, README for OpenSteamworks

Rosentti committed:
new poller system

Rosentti committed:
- login almost done
- slightly better re-translation events
- second auth factors not shown on screen, but you can login
with steam mobile app
- switched all Views to TranslatableInit

Rosentti committed:
login improvements
begin work on logout, mainwindow

Rosentti committed:
logoff work

Rosentti committed:
fix errors preventing startup
add another loggedoff condition

Rosentti committed:
initial layout pass

Rosentti committed:
fixed cmakelists to not create Debug on MSVC...

Rosentti committed:
- refactor
- trayicon translation
- generic avaloniaobject translation

Rosentti committed:
CUtlMap impl

Rosentti committed:
auto-exit serviced and htmlhost on parent death

Rosentti committed:
CUtl improvements

Rosentti committed:
- moved protobufs into separate project
- added webui protobufs
- start working on library code
- first part of sharedconnection rework

Rosentti committed:
CUtlMap: better LessFunc system

Rosentti committed:
fixed cmake.sourceDirectory

Rosentti committed:
- partially implemented cloudconfigstore
- fixed clientmessaging to free connections properly
- start implementing appmanager

EoceneMiacid commented:
Moving away from C++/Qt is a big mistake and frankly, kills my interest in this project stone dead.

Rosentti committed:
cloudconfigstore work
will break your collections however

Rosentti committed:
oops, forgot the Key argument

Rosentti committed:
added ipc blacklist feature
guards against crashing the existing client

Rosentti committed:
workity work

Rosentti committed:
update OSW, start work on integrating collections

Rosentti committed:
- collections
- improved App and AppsManager
- added logger (TODO: vconsole type GUI for it)
- TODO: switch all instances of Console.WriteLine to Logger
- TODO: ui for library
- TODO: remove Component class, split it into:
  - IHasLifetime
  - IHasUserLifetime
  - IRegistersComponents

GermanAizek commented:
> While C# is a better choice to not deal with UB, raw memory and threads, it's still mostly controlled by a single entity (Microsoft) and to me was always a pain to run anything. Development tools for C# on Linux are varying from non-existent to barely functioning.
> 
> Sorry, didn't want to sound like a fossbro, but C# is still a concern. All in all I still agree with you, DepotDownloader for example is a pretty cool program and I used it to get game data on platforms that can't run Steam. I found out about this project today and didn't knew you planned migration to dotnet.

@a1batross in this case, everything is true, you said. Why use C# in linux. It's such an amazing choice.

Rosentti committed:
- refactored container/component system
- improved shutdown logic
- logoff progress system
- new lifetime system
- new loginlifetime system
- nonfunctional chromehtml jankery
- replaced global using with custom value types
- custom value type support for JITEngine
- defined struct packing in SteamClient
- registering callbacks by id
- allow pausing callbackthread
- callresult work started
- added 1280022 callbackname
- changed accessibility to public for structs
- logon now waits for appinfo updates
- PostLogonState_t slightly improved
- updated WarningMessageHook_t
- updated PostAPIResultInProcess
- fixed delegate JITting
- added ISteamClient and ISteamHTMLSurface for testing purposes
- removed bystack returns (unused)
- added ILGeneratorEx, which can print the generated IL for debugging
- added notes on building dotnet to diagnose JIT issues
- proper implementation for generating the same class multiple times
- remote play related stuff
- clarified InProgressLogonSteamID
- improved IsLoggedOn logic
- fixed bad usage of async/await in certain places
- htmlhost now fakes it's executable's location
- assertnotnull shows the correct line number
- added EBrowserType enum in htmlhost
- retyped/renamed some fields of HTMLOptions to maybe be more accurate

Rosentti committed:
remove clientconsole from actions

Rosentti committed:
updated new icon and readme

Rosentti committed:
- initial htmlk browser implementation (buggy)
  - typing keys don't work for some reason
  - it's very pixelated on new avalonia builds
  - crashes on resize
  - hacky x11 key translation
  - typing unsupported on all non-linux os:ses
- working on Avalonia PR to add NativeKeyCode proper
- indev work for convar system
- improved callbackmanager
- removed steamutl as it is unused
- renamed some gitmodules

Rosentti committed:
- htmlhost compiles on windows
  - only as 64-bit, need to rewrite native compilation to fix it
- started rewriting native compilation
- not yet hooking the various funcs

Rosentti committed:
- initial natives rewrite
  - untested on windows and macos
  - osxcross working
  - mingw working
- TODO: make rid-specific builds of OpenSteamworks (Pack size const)

Rosentti committed:
why did the submodules break

Rosentti committed:
fixed

Rosentti committed:
Update README.md

Rosentti committed:
fixed some warnings

Rosentti committed:
- updated blacklist in oswupdater

Rosentti committed:
- updated osw

Rosentti committed:
WIP windows support things

Rosentti committed:
- html works on Windows
- ELanguage helpers moved
- some OSW types
- html options in config file
- ANSI color codes in console
- removed debugger waits

Rosentti committed:
added some readme entries

Rosentti committed:
- in tree avalonia build
- fixed Directory.Build.props case sensitivity
- fixed mingw compile
- fixed linux steamservice

Rosentti committed:
- fix miniutl url

Rosentti committed:
- updated miniutl with fix

Rosentti committed:
- sdl builds
- filter out unwanted files from natives

Rosentti committed:
Delete cibuild.sh

Rosentti committed:
Update dotnet.yaml

Rosentti committed:
Update OpenSteamworks.Native.csproj

Rosentti committed:
remove PublishTrimmed

Rosentti committed:
- logger used almost everywhere now
- sped up startup (somehow)
- makesig script for ghidra
- refcount utility class
- misc cleanups
- attempted to fix cutlbuf resize crash (didn't work)
- split installmanager into it's own class
- fix wrong name in app.manifest

Rosentti committed:
- signature scanning faster than ever
- mprotect correctly
- NativeLibrary utility class
-  CUtlString fixes
- removed C# dev kit from being recommendable by VSC

Rosentti committed:
removed console log

Rosentti committed:
- theoretical windows support
- theoretically fix gh actions
- fix excess language folders in output
- change log file extension to .log to get better highlighting in Kate

Rosentti committed:
- fix windows section names
- remove unneeded code from hookfunction
- exiting will now ignore logout failures
- container logon lifetime index logging fixed

Rosentti committed:
Update README.md

Rosentti committed:
Update README.md

Rosentti committed:
- CGameID rewrite
- renamed some enums
- added a copule appinfos for each app
- more utils for OSW
-

Rosentti committed:
Merge branch 'c#-remake' of github.com:Rosentti/opensteamclient into c#-remake

Rosentti committed:
- fix actions compile

Rosentti committed:
fix avalonia submodule
wtf git...

Rosentti committed:
maybe this will fix it

Rosentti committed:
that didn't work, maybe this will

Rosentti committed:
- this oughta work

Rosentti committed:
remove App class

Rosentti committed:
- hacky custom launch method
- interface debugger supports structs better
- typed some OSW functions
- shuffled around library stuff
- removed usage of value type nullables as they're terrible to work with
- sublogger system
- temporary env vars system
- moved setenvironmentvariable to a better location

Rosentti committed:
(burning) hotfix
- this is dumb

Rosentti committed:
- set default project
- OSW typings

Rosentti committed:
in progressss stuff

Rosentti committed:
- app prototyping
- proton games "work"
- absolutely terrible environ workaround

Rosentti committed:
lots of WIP stuff

Rosentti committed:
WIP shortcuts stuff
fixed compile as well

Rosentti committed:
- cursed native protobuf pointer passing stuff
- fixed html inputs
- interface debugger improvements
- LOTS of types (lots of untested shit as well)
- cell list...
- todo app interface stuff
- utility class for auto resizing buffers
- also some vr stuff as well
- callback logging settings
- bin symlink points to correct place now

Rosentti committed:
- initial tabbing gui
- html surfaces cleanups
- avalonia improvements
- avalonia blur hack fix
- no more graceful quits
- enum renamings
- misc cleanups
- some osw typings
- protobuf hack improvements

Rosentti committed:
- fix htmlsurface logins
- htmlsurface language fixes
- htmlsurface navigation elements
- readme updates
- misc typings

Rosentti committed:
Update dotnet.yaml

try fix gh actions (windows build failing again...)

Rosentti committed:
Update dotnet.yaml

Rosentti committed:
Update dotnet.yaml

Rosentti committed:
- fix fatal error exits
- disable store and community page if webviews disabled

Rosentti committed:
- initial games list view
- TODO: fix resize

Rosentti committed:
wip rewrite of:
- config system
- some bits of library system
- app system
other stuff:
- some OSW types
- new callbacks
- lots of misc cleanups

Rosentti committed:
- some app improvements
- fix enum naming in some places

Rosentti committed:
WIP library assets stuff
- binary vdf cli tool

Rosentti committed:
WIP library stuff

Rosentti committed:
fix for chevron hiding feature

Rosentti committed:
shortcuts work

Rosentti committed:
new steamworks dumper

Rosentti committed:
updated OSW. everything is broken now.

Rosentti committed:
- fixed build
- finally added a steam service sigscanner

Rosentti committed:
updated OSW to beta

Rosentti committed:
Update README.md

Rosentti committed:
- misc OSW typings
- protobuf tester tool
- SDR connection stats
- readd untyped callback support
- improve callbackmanager by making it less spaghettious
- unknown ret 64 and unsigned 64

Rosentti committed:
reversed the clientpipe protocol
TODO: callbacks

Rosentti committed:
ipc improvements
it's a LOT less janky now
dare I say, maybe even production ready...

Rosentti committed:
- switch from concrete SteamClient class to ISteamClient interface
- quick bodge so that the library can now start games with 1 launch option
- library interface that does stuff
- OSW update
- oswupdater dumps:
  - fenceposts, function ids
  - interface ids
  - IPC'd function args and returns
  - blacklisted in cross process data
- include OpenSteamworks.Native in OpenSteamworks instead of OpenSteamworks.Client (needed for protobufhack)
- CSteamID reworked to use properties

Rosentti committed:
osw update

Rosentti committed:
interface debugger fixes

Rosentti committed:
misc changes

Rosentti committed:
working on concommands...

Rosentti committed:
concommands callable now

Rosentti committed:
hacky stuff
- library icons
- concommands ""ui""
- ran a dotnet format on all files
- fixed client never logging on

Rosentti committed:
WIP library assets bodge
does not compile

Rosentti committed:
still doesn't work
WHY IS THE THREAD NOT RUNNING????
AND UpdateLibraryAssets GETS STUCK FOR SOME REASON??????? WTFFFFF

Rosentti committed:
almost works. just need to write our own kvparser

Rosentti committed:
- replaced ValveKeyValue with our own KV parser
  - doesn't parse TM marks correctly
  - somehow causes the asset thread to freeze
  - saving assets.vdf is completely broken now, need to fix TypedKVObject
TEMP STUFF; WILL NOT WORK PROPERLY

Rosentti committed:
- fixed library assets recursive crash
- fixed TM symbols and whatnot
- library assets caching is still broken, now icons no longer load

Rosentti committed:
- fixed caching icons
- caching everything else is still broken
- TODO: fix those
- TODO: filter non-games from the library

Rosentti committed:
- can install games now and pick library folder
- broke the icon caching somehow lol
- STILL TODO: fix library assets caching (not icons)

Rosentti committed:
- rudimentary launching of games with
single option per platform

- fixed some games not showing in library

Rosentti committed:
- remove temp frame saving for htmlsurface
- fix crash on translation failure
- fix library assets caching AND showing
  - logo position is not considered, which leads to bangers such as:
    - life 2 (hl2)
    - fia (mafia)
    - heroes 2 (company of heroes 2)
- possibly fix downloading apps
- do not consider apps with no library assets to be downloadable
- fix language switcher not saving
- fix tool scripts to be runnable from any directory
- add missing finnish translation

Rosentti committed:
- add option to login to friend network on logon
(might not work)

- reworked fatal exception logic, now errors which happen
during logon kill the app

Rosentti committed:
temp dev work for symlink mapping
(should make steamnetworking games work)

Rosentti committed:
- fixed launching some games
- updated app up to date logic
- partially reversed appupdateinfo
- blacklisted steam client and steam from apps
- symlink steamwebrtc library to hopefully get lethal company working
- asset download logic changes:
  - don't try to download empty icons
  - fabricate fake expire dates so downloads don't get retried when they haven't changed
- fixed app logger always showing an invalid appid
- deduplicate kv keys with string intern functionality (ram usage improvement?)
- some OSW typings
- fixed user configs being stored into the root directory instead of the user specific directory
- fixed IDisposable leak in bootstrapper (we should have a tool for this...)
- replaced our reaper with plagman's version (CS2 launches now)
- generate library assets at runtime for apps that are missing them

Rosentti committed:
- update avalonia
- add note about webhelper being broken

Rosentti committed:
-hide some tray icons when not logged in
- fix library sorting
- lots and lots of OSW typings
- download library assets for all games in your collections
- TODO: hide configs from the library always. Will require some rewrites

Rosentti committed:
hack fix for startup crash with library asset gen

Rosentti committed:
fixed the fix

Rosentti committed:
- tried to make  cs2 work to no avail
- fixed launch options stuff
- wrote some friends stuff for later
- some refactorings and whatnot

Rosentti committed:
- fixed launchwrapper
- fixed library asset text logo replacement for some apps
- misc OSW typings

Rosentti committed:
This commit will fix the erroneous updates happening.
(and maybe the windows error popups)
- force disable shader caching always
- hack stuff for windows restarts
- hack stuff for cross process compatmanager
- link libraryfolders.vdf together

Rosentti committed:
Merge branch 'c#-remake' of github.com:Rosentti/opensteamclient into c#-remake

Rosentti committed:
- add stub downloadmanager
- steam cloud stuff
- lots of app and cloud related typings
- removed all references to [IPCOut]
- possibly fixed some funky stuff with dyn collections

Rosentti committed:
updated the placeholder download stuff
possibly fixed terminating games

Rosentti committed:
more downloads stuff

Rosentti committed:
removed viewlocator, as we don't use it

Rosentti committed:
- local package server URL config
- fixed FindControlNested logic
- created an installer for windows
  - GUI isn't finalized
  - lots of TODOs
  - still need to write the actual installation logic

Rosentti committed:
- installer fixes

ThatOneCalculator commented:
Avalonia is great, but unfortunately doesn't adapt to desktop consistency, would you consider making an adapter for GTK4 with something like [Nickvision's Aura](https://github.com/NickvisionApps/Aura) or [GTK4DotNet](https://github.com/uriegel/Gtk4DotNet)? It's an (admittedly borderline unreasonable) tall ask so I'd completely understand if this is out of scope, but having a truly native Steam client for desktop Linux with GTK4 and Libadwaita would be amazing.

Rosentti committed:
compiles and "runs" on Windows
- really shitty atm:
  - doesn't save configs whatsoever
  - long startup time
  - will crash with memory violation errors due to some CUtlBuf shenanigans
  - had to force global.json
  - also had to remove concommands on nonlinux

- fixed the container registration issue
  (how tf did it work before?)

- fixed infinite verifying SDL3 issue
- compile with minsizerel, hope that doesn't break anything

Rosentti committed:
WIP installer stuff

Rosentti committed:
Merge branch 'c#-remake' of github.com:Rosentti/opensteamclient into c#-remake

Rosentti committed:
- windows fixes
- removed clientconsole
- vac support on windows yay
- fixed non-english systems with float parsing
- removed leftover debug /tmp fs accesses

Rosentti committed:
- fixed explicitly added apps for dyn collections

- retranslate when translatable changes

- fixed library assets for apps with parents
- TODO: fix the textblock scaling for apps without logos

Rosentti committed:
fix bootstrapper issue

Rosentti committed:
update avalonia submodule

Rosentti committed:
- port to net8
- remove our custom avalonia fork
- make the html crash more rare
- fix KVObject warning

Rosentti committed:
fix GH actions

Rosentti committed:
updated OSW to depend on our nuget packages instead

Rosentti committed:
remove unused content from this repo

Rosentti committed:
fixup warnings and type conflicts

Rosentti committed:
- new bootstrapper stub
- change bootstrapper init slightly
- remove deleted projects from SLN

Rosentti committed:
almost works. need to figure out how to call managed functions proper

Rosentti committed:
MainExe now runs! yay.
just need to make it bootstrap, update and load the UI

Rosentti committed:
added a couple bootstrapper funcs
will do the actual logic etc tomorrow

Rosentti committed:
WIP mainexe stuff

Rosentti committed:
accomodate for the new natives type

Rosentti committed:
port mainexe to the new native build system

Rosentti committed:
Merge branch 'c#-remake' of github.com:OpenSteamClient/OpenSteamClient into c#-remake

Rosentti committed:
- rename ClientUI to OpenSteamClient
- remove developing guide as it's severely out of date
- remove last remaining traces of ClientConsole
- add AvaloniaCommon to house common Avalonia bits
- document our TODOs in the README
- remove ArtifactsPath from being used for non-publish
- remove MainExe, too much of a headache to be worth it

Rosentti committed:
added some extra goals

Rosentti committed:
rename EAppUpdateError to EAppError

Rosentti committed:
Merge branch 'c#-remake' of github.com:OpenSteamClient/OpenSteamClient into c#-remake

Rosentti committed:
- fixed some possible stuck spots
- finalized rename
- other stuff

Rosentti committed:
- osw update
- include the various OpenSteamworks projects here for easy use of OSWUpdater
- use oldprotoc
- disable dumping offsets
- remove all leftover references to StarInternal, as it was removed

Rosentti committed:
- profiler
- wip shortcuts stuff
  - crashes with tls error
- added (broken) support for return by stack
- throbber logic stuff fix

Rosentti committed:
small update

Rosentti committed:
Merge branch 'c#-remake' of github.com:OpenSteamClient/OpenSteamClient into c#-remake

Rosentti committed:
fix shader system being in a loop
fix bootstrap order

Rosentti committed:
- improve profiler
- fix startup taking long

Rosentti committed:
misc stuff

Rosentti committed:
how does tls work?

Rosentti committed:
assembly hacks, doesn't work

Rosentti committed:
doesn't work, have to commit this so I can update osw

Rosentti committed:
finish osw update

Rosentti committed:
vs code settings

Rosentti committed:
Merge branch 'c#-remake' of github.com:OpenSteamClient/OpenSteamClient into c#-remake

Rosentti committed:
made the ipc client real
- doesn't work yet (will hang on loading apps)
- will eventually replace using native interfaces altogether
- only a single function in clientshortcuts for the time being
removed jitengine support for by stack returns (it was broken)
removed the native test app
removed the "ignoring profile of X on wrong thread"

Rosentti committed:
wip ipc client stuff

Rosentti committed:
fixed the ipc client
- now does reads and writes blocking, like the official client
- removed the callback stuff, as I'm not sure how it works yet
TODO: apps manager breaks now, maybe due to the shortcuts stuff?

Rosentti committed:
- fix app cache
- fix library asset margins

Rosentti committed:
update advanced config to always save

Rosentti committed:
various native fixups

Rosentti committed:
Update README.md

Rosentti committed:
fix typo

Rosentti committed:
- add contributing guidelines
- improve readme
- delete some useless stuff
- various fixes
- remove CB support from IPCClient as it's terribly broken
- fix sharedconnection
- tried to fix library, but the whole thing needs a rewrite eventually
- possibly fixed cloud config store
- half implemented sourcemods
- don't recreate steam.pid if there's already a steam process running
- fixed another login hang issue
- fixed translatables not updating with bindings
probably other gems in here as well, but it's been such a long time since the last update, I've forgotten most of the changes i've done here

Rosentti committed:
Merge branch 'c#-remake' of github.com:OpenSteamClient/OpenSteamClient into c#-remake

Rosentti committed:
semi large update
- library improvements:
  - don't show empty collections
  - search bar
  - live collections
  - launch option picker
- concommands system
  - convars are still unexplored territory
- create our custom observablecollection
- removed ipcclient
  - we need to find another way to do shortcuts stuff
  - ipcclient is too slow and not cross platform
- fixed nativeclient logs
- JITEngine:
  - properties/fields support
  - get ptr of generated class
  - fix generating the same class with a different pointer
- async logger
- disable steamhtml and all web pages
- fix some erroneous logging statements
- localized tab headers

Rosentti committed:
various fixups

Rosentti committed:
OSW update

Rosentti committed:
- fix intermittent hang issue with async logger
- partially add the new timeline API
- remove a bit of extra space before app elements in library app list

Rosentti committed:
handle new runtimes, fix launch

Rosentti committed:
oops.

Rosentti committed:
- friends list icons and invites
- add supported OS list
- observablecollection fixes

Rosentti committed:
runtime bootstrapping issue debugging

Rosentti committed:
library fix

Rosentti committed:
- overlay executable WIP
- friends list localizations
- misc friends list improvements
- fix search bar
- fix collections
- fix actions builds by copying files from the current dir
  - now need to specify all files to copy explicitly instead

Rosentti committed:
fix steamserviced launch issue

Rosentti committed:
another round of bugfixes

Rosentti committed:
flush the logger's filestream on each line

Rosentti committed:
fix library filters

Rosentti committed:
run ourselves in the 32-bit runtime

Rosentti committed:
misc changes
- logoff progress
- download queue is now semi usable
- overlayui stuff

Rosentti committed:
overlayui desktop

Rosentti committed:
fixup some files

Rosentti committed:
various fixes and improvements:
- arch PKGBUILD
- family sharing fix
- default compat tools should now apply
- settings skeleton

Rosentti committed:
fix download progress

Rosentti committed:
settings window:
- locale switcher
- global proton settings
- library folder management

Rosentti committed:
- hopefully fix app sections not existing after logon
- removed excessive download logging

Rosentti committed:
oops, accidentally overwrote lots of compat tools with the wrong priority setting

Rosentti committed:
- overlay debug target
- overlay IPC is basically finished
- overlay inputs seem to work OK, keyboard untested
now I "just" need to write the native parts and do some fixups

Rosentti committed:
some more WIP overlay stuff

Rosentti committed:
OSW update
- fix for new utlmemory ABI
- IClientBluetoothManager got deleted
- heavy runtime was removed

Rosentti committed:
OSW update

Rosentti committed:
minor improvements:
- remove some leftover usage of Console.WriteLine
- transport support (if it's helpful in any way)

Rosentti committed:
- reversed the clientsocket thingy
- made some app launch failures non-fatal

Rosentti committed:
- re-introduce MainExe
- remove bootstrappership preload
- fix steamservice failing to init (serviced was fine, the local client failed)

Rosentti committed:
- new webhelper system
- re-enabled the web pages
- possibly fixed GH actions

Rosentti committed:
remove remaining traces of htmlhost

Rosentti committed:
Update README.md

Rosentti committed:
chrome stuff

Rosentti committed:
misc fixes, add goals for the eventual looming rewrite

Rosentti committed:
Merge branch 'c#-remake' of github.com:OpenSteamClient/OpenSteamClient into c#-remake

Rosentti committed:
OSW update

Rosentti committed:
HUGE refactor. do not update to this version, it's severely undercooked ATM

Rosentti committed:
nativeaot fixups round 2

Rosentti committed:
fix bool callbacks and logger newlines

Rosentti committed:
- fix enumerable tostring with callbackmetadata
- protobuf non-ptr
- sharedconnection rewrite
- remaining OSW nativeaot fixes
- ILoggerFactory usage

Rosentti committed:
update docs, actions, and PKGBUILD

Rosentti committed:
yaml uses spaces?

Rosentti committed:
this should do

Rosentti committed:
actions fix 3

Rosentti committed:
update these to v4

Rosentti committed:
library redesign goals

Rosentti committed:
build from OSW nuget

Rosentti committed:
update to the new OSW version

Rosentti committed:
build from NuGet.org :)

Rosentti committed:
Update Directory.Build.props

Rosentti committed:
run actions when changing Directory.Build.props

This allows for really simple OSW updates.

Rosentti committed:
fix versions

Rosentti committed:
Merge branch 'c#-remake' of github.com:OpenSteamClient/OpenSteamClient into c#-remake

Rosentti committed:
fix wrong version

Rosentti committed:
made it work again:
- refactored the app system (unsuccessfully)
- add editorconfig
- adapted to new OSW changes
- yeah

Rosentti committed:
fix actions

Rosentti committed:
misc changes:
  - compat tool selection per apps
  - delete unused dumped_data folder (this resides in OSW nowadays)
  - fix configs and other stuff being visible in library
  - exit when X pressed on loading dialog
  - collapse library to list only, like in original OSC
  -

Rosentti committed:
slight improvements:
 - color uninstalled games grey
 - added more settings to app settings panel
